### PR TITLE
Added unit tests

### PR DESCRIPTION
**Pull request by antonag32**
Added unit tests as asked by Moises. Skipped one of the cases (creating a session and following the workflow) as there is no session workflow in this version.

### CODE REVIEW

**Comment by antonag32** [comment]
@frahikLV Please review my changes. Thanks.

File: opa/tests/test_opa.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+from odoo.tools import mute_logger
+
+
+class TestOpa(TransactionCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        opa_manager_group = cls.env.ref('opa.group_manager')
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: Use `browse_ref` to get the object or `ref` to get the id

I suggest:
```suggestion
        opa_manager_group_id = cls.ref('opa.group_manager')
```

File: opa/tests/test_opa.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        cls.responsible_user = cls.env['res.users'].create({
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: Move this to demo/res_users.xml

File: opa/tests/test_opa.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            'name': 'Mr. Responsible',
+            'login': 'mresponsible@vauxoo.com',
+            'email': 'mresponsible@vauxoo.com',
+            'groups_id': [Command.set([opa_manager_group.id])]
+        })
+
+        cls.session_instructor = cls.env['res.partner'].create({
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: Move this to demo/res_partner.xml

File: opa/tests/test_opa.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            'name': 'Super Instructor',
+            'email': 'best.instructor@vauxoo.com',
+            'instructor': True
+        })
+
+        cls.course_1_title = 'How not to do Unit Tests'
+        cls.course_1 = cls.env['course'].create({
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: Move this to demo/course.xml

File: opa/tests/test_opa.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            'title': cls.course_1_title,
+            'description': 'Ft. this repo',
+            'responsible_id': cls.responsible_user.id
+        })
+
+    def test_session_instructor_not_attendee(self):
+        """ Validate that a session's instructor can't be in the attendees list. """
+        with self.assertRaises(ValidationError):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: change this to 
```suggestion
        with self.assertRaisesRegex(ValidationError, "The msg error that you expect"):
```

File: opa/tests/test_opa.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            self.env['session'].create({
+                'title': 'Instructor teaches itself',
+                'course_id': self.course_1.id,
+                'starts_on': fields.Date.today(),
+                'instructor_id': self.session_instructor.id,
+                'attendee_ids': [fields.Command.set([self.session_instructor.id])]
+            })
+
+        with self.assertRaises(ValidationError):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: same here

File: opa/tests/test_opa.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                'title': 'Sneaky Instructor',
+                'course_id': self.course_1.id,
+                'starts_on': fields.Date.today(),
+                'instructor_id': self.session_instructor.id
+            })
+            test_session.write({
+                'attendee_ids': [fields.Command.set([self.session_instructor.id])]
+            })
+
+        with self.assertRaises(ValidationError):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: same here

File: opa/tests/test_opa.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            self.env['session'].create({
+                'title': 'Instructor teaches itself',
+                'course_id': self.course_1.id,
+                'starts_on': fields.Date.today(),
+                'instructor_id': self.session_instructor.id,
+                'attendee_ids': [fields.Command.set([self.session_instructor.id])]
+            })
+
+        with self.assertRaises(ValidationError):
+            test_session = self.env['session'].create({
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: move this part outside the assert

File: opa/tests/test_opa.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                'course_id': self.course_1.id,
+                'starts_on': fields.Date.today(),
+                'instructor_id': self.session_instructor.id
+            })
+            test_session.write({
+                'attendee_ids': [fields.Command.set([self.session_instructor.id])]
+            })
+
+        with self.assertRaises(ValidationError):
+            test_session = self.env['session'].create({
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: move this part outside the assert

File: opa/tests/test_opa.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                'attendee_ids': [fields.Command.set([self.session_instructor.id])]
+            })
+            test_session.write({
+                'instructor_id': self.session_instructor.id
+            })
+
+    def test_session_required_course(self):
+        """ Ensure a course is required when creating a session. """
+        with mute_logger('odoo.sql_db'):
+            with self.assertRaises(NotNullViolation):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: use `assertRaisesRegex` instead

File: opa/tests/test_opa.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+                self.env['session'].create({
+                    'title': 'Incorrectly creating sessions',
+                    'starts_on': fields.Date.today()
+                })
+
+    def test_course_same_title_description(self):
+        """ Validate that courses can't have their description be the same as their title. """
+        test_string = 'This is about to blow up!'
+        with mute_logger('odoo.sql_db'):
+            with self.assertRaises(CheckViolation):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: use `assertRaisesRegex` instead

File: opa/tests/test_opa.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                self.env['course'].create({
+                    'title': test_string,
+                    'description': test_string
+                })
+
+    def test_course_repeated_title(self):
+        """ Ensure courses must have unique titles. """
+        with mute_logger('odoo.sql_db'):
+            with self.assertRaises(UniqueViolation):
```
Comment (frahikLV) [inline opa/tests/test_opa.py]: use `assertRaisesRegex` instead

File: opa/tests/test_opa.py (hunk 1 above)
Comment (frahikLV) [reply opa/tests/test_opa.py]: and change the references of `opa_manager_group.id ` to `opa_manager_group_id`

**Review by frahikLV — CHANGES_REQUESTED** [review]
@antonag32 can you address these changes?

**Comment by antonag32** [comment]
@frahikLV 
Hi, I did not use demo data because [according to Odoo:](https://www.odoo.com/documentation/15.0/developer/howtos/rdtraining/E_unittest.html#writing-a-test)

> The tests should be independent from the data currently in the database (including demo data)

I checked some [source code from odoo](https://github.com/odoo/odoo/blob/43fb1c47b4867ac9b1fe7c3dd30e8bed7bd93f13/addons/purchase/tests/test_access_rights.py#L29) and I also found they setup all their test data separately and don't use demo data.

Should I still move the data to demo? I think it goes against the principles of unit test since unit tests should have as little dependencies as possible. If I move it to demo now my tests would depend on the state of the database.

**Comment by frahikLV** [comment]
@antonag32 Don't say that don't use demo data, there is a lot of examples on test that [uses demo](https://github.com/odoo/odoo/blob/bc131c0cfb51c953de8ec41fb820c8c7831eefb5/addons/account_check_printing/tests/test_print_check.py#L24) on odoo community, By independent, means that your tests do not only work with demo. I don't know if this makes sense to you.

On the other hand, when a client module becomes quite large, having it in records, makes it possible to reuse records, and save that creation time, in addition to reducing the code to be used in the tests, making more visible the part that is being evaluated and not doing one hundred lines of code to Test a single method.

Please apply all changes

**Comment by antonag32** [comment]
@frahikLV 
All changes have been made. However I was not able to use `cls.ref('xmlid')` directly. The only way I was able to use it was with `cls().ref('xmlid')` which I'm not sure it is what you wanted. Therefore I used `cls.env.ref`. Please review my changes and tell me how to proceed. Thanks.

**Review by frahikLV — APPROVED** [review]
@antonag32 LGTM +1, I will merge it

**Pull request merged**